MUST for multi-publisher relay handling (#1115)
moq-wg/moq-transport: draft-ietf-moq-transport.md

{{object-properties}}, with the exception of any extensions
({{object-extensions}}) that specify otherwise.

## Multiple Publishers

A Relay can receive PUBLISH_NAMESPACE for the same Track Namespace or PUBLISH
messages for the same Track from multiple publishers.  The following sections
explain how Relays maintain subscriptions to all available publishers for a
given Track.

There is no specified limit to the number of publishers of a Track Namespace or
Track.  An implementation can use mechanisms such as PUBLISH_ERROR,

> PUBLISH_NAMESPACE_ERROR, UNSUBSCRIBE, PUBLISH_NAMESPACE_CANCEL or GOAWAY if it

Author: This comment is about an older version of the text, I already removed GOAWAY.

{{object-properties}}, with the exception of any extensions
({{object-extensions}}) that specify otherwise.

## Multiple Publishers

A Relay can receive PUBLISH_NAMESPACE for the same Track Namespace or PUBLISH
messages for the same Track from multiple publishers and MUST handle them

> uniformly.

Reviewer: I've no idea what "MUST handle them uniformly" means here, especially as a normative requirement.

Reviewer 2: how about "A Relay can receive PUBLISH_NAMESPACE or PUBLISH messages for the same Track Namespace or Track from multiple publishers and MUST process them consistently."

Reviewer 3: I would simply discard this "handle them uniformly" part of the sentence. I don't see reasons a Relay would not act "uniformly", so I don't understand why we would need to mention anything here.

Reviewer 3: ... by the way, what about Publishers who have different publisher_priority? To handle the multiple track _uniformly_, should the Relay force all Publisher to have the same priority by setting the same subscriber_priority to all subscription?

> messages for the same Track from multiple publishers and is expected to treat
> them uniformly.

Author: ```suggestion
messages for the same Track from multiple publishers and MUST handle them
uniformly.
```

> This behavior is necessary to support the following scenarios:

Reviewer 2: I feel adding these use-cases might be causing confusion and perhaps keep this PR scoped just to relay behavior alone?

Author: I think getting agreement on the problems we are trying to solve is important, let's try work through them a bit more.  If we stall out here and the normative parts of this PR have consensus, we can remove this.

Reviewer 3: As said in another comment, I am not a fan of the "MUST handle it uniformly". Similarly, I would not mention "the behavior is necessary" (because in my mind it is just the normal behavior). I would rephrase it such as "Multi-publisher can happen in many scenarios, including (but not restricted to)"

Reviewer 2: > Multi-publisher can happen in many scenarios, including (but not restricted to)"

I think this is a good suggestion

1. A publisher loses state and reconnects to the same Relay, but the Relay

>    has not yet detected the old session has terminated.

Reviewer: I am not sure this is a "multiple publisher" case, as at any individual instance in time, there exists only one publisher that is active?

2. A publisher is performing a graceful migration to a new session.
3. A system that facilitates failover with multiple redundant publishers that

>    produce identical Track content.

Reviewer: 3 is actually two different use cases: one (3a) in which there are two redundant publishers sending the same media, and one (3b) where one of them is on standby.

In all of those cases, there is only one original publisher, but it publishes its track multiple times.  2 and 3b are similar in that most of the time, there is only one publisher speaking, so we might want to add signaling to distinguish this from 3a (which is a use case which we've not discussed that much in the working group, but probably should).

> 4. A distributed set of endpoints, each publishing different content in the same

Author: ```suggestion
4. A distributed set of endpoints, each publishing different Objects in the same
```

>    Track.

Reviewer: This is the only use case that appears to imply two original publishers with disjoint state active at the same time.  I don't believe we've ever had consensus to support this.

Author: Agree it's still marked as Needs Discussion following the consensus call in the draft where we relaxed the temporal requirement for groups.  I will ask the chairs to help clarify this.  I think it matters less for the normative language in this PR, because the same rules work for _any_ multi-publisher SUBSCRIBE case, than in #1117.

Reviewer 4: We have discussed this at several face to face meetings. We implemented this and I don't see any reason it should have any negative impact on scalability.

Relays MUST handle Objects for the same Track from multiple publishers and
forward them to active matching subscriptions. The Relay SHOULD attempt to

> deduplicate Objects before forwarding, subject to implementation constraints.

Reviewer 3: In the case the Track is such as group ID increases (and possibly sequentially), it could be expected the Relay to (try to) re-order groups that arrive out of order. 
Of course, I know it can be a dangerous process and I don't want to enter into a very complicated debate here... but...
Should we say something like "_The Relay MAY forward Objects from different publishers in an order that is not the same as the reception order._"?

Author: @gwendalsimon We already constrain delivery order within the same track to be in the specified Group Order (which can differ from reception order) in the Priority section.  So I think we're good?

{{object-properties}}, with the exception of any extensions
({{object-extensions}}) that specify otherwise.

## Multiple Publishers

> A Relay can receive PUBLISH_NAMESPACE for the same Track Namespace or PUBLISH

Author: It can only get a SUBSCRIBE_OK once it issues the SUBSCRIBE, which is after the PUBLISH_NAMESPACE part happens.  There's separate text below that says when it needs to issue multiple SUBSCRIBES.

> A Relay can receive PUBLISH_NAMESPACE for the same Track Namespace or PUBLISH
> messages for the same Track from multiple publishers and is expected to treat
> them uniformly.

Reviewer 2: ```suggestion
A Relay can receive PUBLISH_NAMESPACE or PUBLISH messages for the same Track Namespace or Track from multiple publishers and must handle them uniformly.
```

> The is required in order to support the following use cases:

Reviewer 2: ```suggestion
This behavior is necessary to support the following scenarios:
```

1. A publisher loses state and reconnects to the same Relay, but the Relay
   has not yet detected the old session has terminated.

> 2. A publisher is attempting to gracefully migrate to a new session.

Reviewer 2: ```suggestion
2. A publisher is performing a graceful migration to a new session.
```

3. A system that facilitates failover with multiple redundant publishers that
   produce identical Track content.
4. A distributed set of endpoints, each publishing different content in the same
   Track.

> Relays MUST be able to process Objects for the same Track from multiple
> publishers and forward Objects to active matching subscriptions.  The Relay
> SHOULD attempt to deduplicate Objects when forwarding downstream, subject to
> implementation limits.

Reviewer 2: ```suggestion
Relays MUST handle Objects for the same Track from multiple publishers and forward them to active matching subscriptions. The Relay SHOULD attempt to deduplicate Objects before forwarding, within the constraints of implementation limits.
```

MUST withhold sending FETCH_OK until it does.

For successful subscriptions, the publisher maintains a list of
subscribers for each Track. Each new Object belonging to the

> Track within the subscription range is forwarded to each active

Author: I hear your feedback.  I got a little over-zealous here. I'll try to constrain the mini-corrections to paragraphs that are actually changing.

that have a namespace prefix match.

When a relay receives an incoming SUBSCRIBE that triggers an upstream
subscription, it MUST send a SUBSCRIBE request to each publisher that has
published the subscription's namespace or prefix thereof, unless it already has
an active subscription for the Track requested by the incoming SUBSCRIBE request
from all available publishers.  If it already has a matching upstream
subscription in Forward State=0, it SHOULD send a SUBSCRIBE_UDPATE with
Forward=1 to all publishers.

When a relay receives an incoming PUBLISH message, it MUST send a PUBLISH
request to each subscriber that has subscribed (via SUBSCRIBE_NAMESPACE)
to the Track's namespace or prefix thereof.

> When a relay receives an incoming PUBLISH_NAMESPACE for a namespace that matches
> one or more existing subscriptions to other upstream sessions, it MUST send a
> SUBSCRIBE to the publisher that sent the PUBLISH_NAMESPACE for each matching
> subscription.  When it receives an incoming PUBLISH message for a Track that has
> active subscribers, it MUST respond with PUBLISH_OK, and SHOULD use Forward
> State=1.

Author: ```suggestion
When a relay receives an authorized PUBLISH_NAMESPACE for a namespace that
matches one or more existing subscriptions to other upstream sessions, it MUST
send a SUBSCRIBE to the publisher that sent the PUBLISH_NAMESPACE for each
matching subscription.  When it receives an authorized PUBLISH message for a
Track that has active subscribers, it MUST respond with PUBLISH_OK.  If at least
one downstream subscriber for the Track has Forward State=1, the Relay MUST use
Forward State=1 in the reply.
```

all publishers and PUBLISH_NAMESPACE and PUBLISH messages to all subscribers
that have a namespace prefix match.

When a Relay receives an authorized SUBSCRIBE for a Track with one or more
active upstream subscriptions, it MUST reply with SUBSCRIBE_OK.  If the
SUBSCRIBE has Forward State=1 and the upstream subscriptions are in Forward
State=0, the Relay MUST send SUBSCRIBE_UPDATE with Forward=1 to all publishers.
If there are no active upstream subscriptions for the requested Track, the Relay
MUST send a SUBSCRIBE request to each publisher that has published the
subscription's namespace or prefix thereof.  If the SUBSCRIBE has Forward

> State=1, then the Relay MUST use Foward=1 when subscribing upstream.

Reviewer 2: ```suggestion
State=1, then the Relay MUST use Forward=1 when subscribing upstream.
```

When a relay receives an incoming PUBLISH message, it MUST send a PUBLISH
request to each subscriber that has subscribed (via SUBSCRIBE_NAMESPACE)
to the Track's namespace or prefix thereof.

When a relay receives an authorized PUBLISH_NAMESPACE for a namespace that
matches one or more existing subscriptions to other upstream sessions, it MUST
send a SUBSCRIBE to the publisher that sent the PUBLISH_NAMESPACE for each
matching subscription.  When it receives an authorized PUBLISH message for a
Track that has active subscribers, it MUST respond with PUBLISH_OK.  If at least
one downstream subscriber for the Track has Forward State=1, the Relay MUST use

> Forward State=1 in the reply.

Reviewer 2: This is enforcing a behavior, where when a relay gets PUBLISH with forward=1 and there are no active subscriptions,  a Relay MUST respond with forward=0. This defeats the purpose of publish with forward=1.

Reviewer 2: That seems fine to me per our discussions of making it mandatory only when there are active subscriptions.

that have a namespace prefix match.

When a relay receives an incoming SUBSCRIBE that triggers an upstream
subscription, it MUST send a SUBSCRIBE request to each publisher that has
published the subscription's namespace or prefix thereof, unless it already has
an active subscription for the Track requested by the incoming SUBSCRIBE request
from all available publishers.  If it already has a matching upstream
subscription in Forward State=0, it SHOULD send a SUBSCRIBE_UDPATE with
Forward=1 to all publishers.

When a relay receives an incoming PUBLISH message, it MUST send a PUBLISH
request to each subscriber that has subscribed (via SUBSCRIBE_NAMESPACE)
to the Track's namespace or prefix thereof.

When a relay receives an incoming PUBLISH_NAMESPACE for a namespace that matches
one or more existing subscriptions to other upstream sessions, it MUST send a
SUBSCRIBE to the publisher that sent the PUBLISH_NAMESPACE for each matching
subscription.  When it receives an incoming PUBLISH message for a Track that has

> active subscribers, it MUST respond with PUBLISH_OK, and SHOULD use Forward

Author: Ah yes.  If the downstream forward state=1 for any subscriber then it MUST be 1 upstream.  If all downstream subscribers have forward state=0, I think elsewhere we say SHOULD send 1 upstream but offer the relay some flexibility (for example, DoS prevention).
